### ISSUE TITLE
errcatch fails to catch various errors

### ISSUE THREAD

**Issue by rtoy** [rtoy/maxima#1307]
Imported from SourceForge on 2024-07-04 08:54:22
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2015-03-17 17:21:21
Original: https://sourceforge.net/p/maxima/bugs/2921

---

errcatch doesn't catch a syntax error. e.g.

~~~~
errcatch (parse_string (")abc"));
~~~~

prints an error message but doesn't yield [] as expected. Probably what's going on is that errcatch catches some specific errors but doesn't catch whatever MREAD-SYNERR is throwing (namely whatever is thrown by THROW-MACSYMA-TOP).

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-04 08:54:24
Created by **[sikkenogetvrovl](https://sourceforge.net/u/sikkenogetvrovl/)** on 2017-09-07 13:55:54
Original: https://sourceforge.net/p/maxima/bugs/2921/#177b

---

I am struggeling with the same bug at the moment e.g. :
errcatch(parse_string("Heading="));

Does not give []

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-04 08:54:28
Created by **[kjak](https://sourceforge.net/u/kjak/)** on 2021-10-22 21:03:42
Original: https://sourceforge.net/p/maxima/bugs/2921/#55ac

---

- **summary**: errcatch doesn&#39;t catch a syntax error --> errcatch fails to catch various errors
- **assigned_to**: Kris Katterjohn

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-04 08:54:31
Created by **[kjak](https://sourceforge.net/u/kjak/)** on 2021-10-22 21:03:42
Original: https://sourceforge.net/p/maxima/bugs/2921/#e8d4

---

I just looked at this a bit.

`errcatch` needs to catch throws to `MACSYMA-QUIT` to catch the mentioned parser errors, as well as some other errors (e.g., from the Galois field functions in numth).

It looks like `errcatch` should also catch throws to `ERRORSW` as well.  The `ERRORSW` variable controls whether or not some errors just cause a throw to `ERRORSW` instead of the signaling of a Maxima error.

`ERRORSW` is bound to `T` in various places in src and there are various catchers setup in src.  `ERRORSW` is also bound to `T` while evaluating `errcatch` forms in the hyperint contrib package.

Currently if something in an `errcatch` throws to `ERRORSW` and nothing ever catches it, then `errcatch` will "catch" this due to the eventual lisp error from a throw to a nonexistent tag (and it will also by default print the error report message about the nonexistent tag from lisp and store that message in Maxima's `error` variable).  If a catcher is setup outside the `errcatch`, then `errcatch` will not catch the error.  (`errcatch` should catch this just like it will catch throws to `RAT-ERR` even if something outside the `errcatch` would have caught it.)

There are other tags that `errcatch` *could* catch, but I don't see any other particularly reasonable ones yet.  I don't want to set up a bunch of catchers in `errcatch`, but I'll still look a bit more.

Anyway, I'll update `errcatch` to catch these new "errcatch tags".

In the cases where `errcatch` catches a throw to `MACSYMA-QUIT` or `ERRORSW`, `errcatch` has no info on the error that occurred.  In these cases I'll make sure some message is stored in Maxima's `error` variable stating that `errcatch` caught an error (so `error` doesn't have "No error." or some previous error message in it.)

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-04 08:54:35
Created by **[macrakis](https://sourceforge.net/u/macrakis/)** on 2021-10-22 21:41:00
Original: https://sourceforge.net/p/maxima/bugs/2921/#67d7

---

``parse_string`` also gives very peculiar errors:
~~~~
 parse_string("++");

incorrect syntax: Premature termination of input at $.
rse_string("++");++$
~~~~
Why on earth is ``rse_string("++";`` part of the error?

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-04 08:54:38
Created by **[kjak](https://sourceforge.net/u/kjak/)** on 2021-10-22 22:55:57
Original: https://sourceforge.net/p/maxima/bugs/2921/#67d7/6699

---

From glancing at `MREAD-SYNERR` I'm guessing that `COLUMN` is going back too far in `*PARSE-WINDOW*`.  I'll try to also look at that more closely soon too.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-04 08:54:42
Created by **[kjak](https://sourceforge.net/u/kjak/)** on 2021-10-24 20:22:44
Original: https://sourceforge.net/p/maxima/bugs/2921/#fb45

---

- **status**: open --> closed

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-04 08:54:45
Created by **[kjak](https://sourceforge.net/u/kjak/)** on 2021-10-24 20:22:44
Original: https://sourceforge.net/p/maxima/bugs/2921/#8670

---

The bug where `errcatch` failed to catch some errors has been fixed by commit [f11cab].  I just stuck with `ERRORSW` and `MACSYMA-QUIT` as I described before.

```
(%i2) errcatch (parse_string ("++"));

incorrect syntax: Premature termination of input at $.
e_string ("++"));++$
                  ^
(%o2) []

(%i3) errcatch (gf_set_data ("wtf"));
`gf_set_data': Field characteristic must be a prime number.
 -- an error.
(%o3) []

(%i4) block ([?errorsw : true], errcatch (log (0)));
(%o4) []
```

Errors from things like `%i2` and `%i3` were uncaught by `errcatch` before.  For something like `%i4` it was like I described in my previous comment about `ERRORSW`.


For the other issue that Stavros mentioned regarding error messages, I'll need to look into all of this more because there is another issue.  I think my initial guess about `COLUMN` was correct, but when I started testing I managed to get a lisp error from a bogus comparison (without me changing anything).  I don't feel like messing with that right now, but I've made note of these issues and I'll try to look into them further soonish.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-04 08:54:49
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2021-11-01 01:34:57
Original: https://sourceforge.net/p/maxima/bugs/2921/#b725

---

Kris, thanks for working on this problem. For the record, I worked on a similar problem some time ago, and what was able to figure out is in:
https://github.com/maxima-project-on-github/maxima-packages/tree/master/robert-dodier/catchable_syntax_error

You mentioned that `error` doesn't have a helpful error message afterwards -- I guess I was able to figure out how to put the parser error message into `error`. It is a minor point.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-04 08:54:52
Created by **[kjak](https://sourceforge.net/u/kjak/)** on 2021-11-01 21:21:28
Original: https://sourceforge.net/p/maxima/bugs/2921/#b725/e7ab

---

Hi Robert.  Thanks for the comment and the link.  Sorry I didn't remember your previous work on that.

Yes, the way I went about this was focusing on `errcatch` since there were other errors that weren't being caught.

My initial idea was to focus on `MREAD-SYNERR` itself, which is what you did.  I see that your new `MREAD-SYNERR` calls `error`, which calls `MERROR`, and so the `error` variable gets updated with the actual error message.  (In my approach I also call `MERROR` to get `error` updated, but I only have a generic error message since I call from `errcatch`.)

I notice that you do not have stuff like line info in your error messages anymore (which you also mention in a comment).  My original idea for changing `MREAD-SYNERR` was to basically build up a string with the existing info and pass that to `MERROR` so that the `error` variable would be updated and `errcatch` would catch the error.  I didn't think about that much and I didn't actually try that before moving on to `errcatch` itself.  I wonder if things like line info are missing from yours because you tried something like that and ran into problems.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-04 08:54:56
Created by **[kjak](https://sourceforge.net/u/kjak/)** on 2021-11-01 21:34:02
Original: https://sourceforge.net/p/maxima/bugs/2921/#e02f

---

Last night I was able to look into the issue that Stavros mentioned and the lisp error I mentioned before.

I have fixes for both: one involving changing a binding in `PARSE-STRING` so that `*PARSE-WINDOW*` is updated appropriately in `MREAD` and the other involves checking for the EOF value in when traversing `*PARSE-WINDOW*` (this EOF value can come from loading a Maxima file, for example).  The lack of EOF checking was causing the bogus comparison I mentioned.

After that I found another lisp error I need to investigate.  I hope to have time to look at this more later this week.
